### DISCUSSION TITLE
docker orchardcore-cms-linux localization language invalid

### DISCUSSION THREAD

**Discussion by snmin** [OrchardCMS/OrchardCore#11437] [Q&A]
The localized language of docker orchardcore-cms-linux is invalid. It is possible to try to run the localized language with Visual Studio under windows.

**Comment by hishamco** [comment]
@Skrypt I think you are the one who worked in this before, any idea?

**Reply by Skrypt** [reply]
It was an issue with the CI that we fixed. Of course, some Docker images are still missing these but the latest ones should not.

**Reply by Skrypt** [reply]
I just tested with this command and it works perfectly : 

```cmd
docker run --name orchardcms -p 8080:80 orchardproject/orchardcore-cms-linux:latest
```

https://user-images.githubusercontent.com/3228637/160428780-38a349eb-32f7-4079-b4ec-52279cecd2c5.mp4

**Reply by snmin** [reply]
Yes, I do have the option to select the localized language in the installation interface, but the localized language selection will be invalid after installation.

**Comment by snmin** [comment]
Hi, I just tested it again. Used the command you just gave me. The problem is still the same. Setting the localization language is invalid. My browser is chrome and I also tested it with Microsoft Edge. In addition, it will be displayed incompletely under the IE browser, and the options cannot be displayed when clicking on the user name.


IE:


***@***.***
 
From: Jasmin Savard
Date: 2022-03-28 23:05
To: OrchardCMS/OrchardCore
CC: snmin; Author
Subject: Re: [OrchardCMS/OrchardCore] docker orchardcore-cms-linux localization language invalid (Discussion #11437)
I just tested with this command and it works perfectly :
docker run --name orchardcms -p 8080:80 orchardproject/orchardcore-cms-linux:latest

**Reply by snmin** [reply]
![image](https://user-images.githubusercontent.com/44972156/160441660-c79a1c09-5698-479e-b6ec-b8f193b39426.png)
![image](https://user-images.githubusercontent.com/44972156/160441685-1cdec5a1-a790-4b85-9edc-c6faed2892e2.png)

**Comment by snmin** [comment]
For this problem, I used VS to redistribute the linux package, and it is the same when running under ubuntu.



***@***.***
 
From: Jasmin Savard
Date: 2022-03-28 23:05
To: OrchardCMS/OrchardCore
CC: snmin; Author
Subject: Re: [OrchardCMS/OrchardCore] docker orchardcore-cms-linux localization language invalid (Discussion #11437)
I just tested with this command and it works perfectly :
docker run --name orchardcms -p 8080:80 orchardproject/orchardcore-cms-linux:latest

**Reply by Skrypt** [reply]
There's probably something missing in your Linux distro. 

Make sure `ICU library` is installed.

```cmd
sudo apt-get install -y icu-devtools
```

**Comment by snmin** [comment]
Yes I'm sure I have icu-devtools installed





***@***.***
 
From: Jasmin Savard
Date: 2022-03-28 23:38
To: OrchardCMS/OrchardCore
CC: snmin; Author
Subject: Re: [OrchardCMS/OrchardCore] docker orchardcore-cms-linux localization language invalid (Discussion #11437)
There's probably something missing in your Linux distro.
Make sure ICU Lib is installed.
sudo apt-get install -y icu-devtools

**Reply by Skrypt** [reply]
Then which culture are you trying to use? If none of the cultures are working then I might ask if you are using the same Docker image as I did for testing?

**Comment by snmin** [comment]
I want to use Chinese zh.

thank you


发送自我的三星盖乐世智能手机。


-------- 原始信息 --------
发件人： Jasmin Savard ***@***.***>
日期: 2022/3/29 00:54 (GMT+08:00)
收件人： OrchardCMS/OrchardCore ***@***.***>
抄送： snmin ***@***.***>, Author ***@***.***>
主题： Re: [OrchardCMS/OrchardCore] docker orchardcore-cms-linux localization language invalid (Discussion #11437)


Then which culture are you trying to use?

―

**Reply by Skrypt** [reply]
Please take a look at : https://github.com/OrchardCMS/OrchardCore/issues/5595

**Comment by snmin** [comment]
OK, how about tomorrow? It's already more than 1 a.m. on my side. I have to go to bed. If you can, I can give you the IP address and management information of the server.



发送自我的三星盖乐世智能手机。


-------- 原始信息 --------
发件人： Jasmin Savard ***@***.***>
日期: 2022/3/29 00:59 (GMT+08:00)
收件人： OrchardCMS/OrchardCore ***@***.***>
抄送： snmin ***@***.***>, Author ***@***.***>
主题： Re: [OrchardCMS/OrchardCore] docker orchardcore-cms-linux localization language invalid (Discussion #11437)


Please take a look at : #5595<https://github.com/OrchardCMS/OrchardCore/issues/5595>

―

**Reply by Skrypt** [reply]
There is an issue with our Crowdin translations for Chinese translations. Mainly because we have 2 different cultures here and we have nothing for "zh" as a fallback. 

So, what currently works is zh-CN or zh-TW; but these are not present under Linux. They are zh-Hans-CN and zh-Hans and probably zh-Hant.
Renaming these folders may fix the issue but every time you will rebuild it may remove your custom folders.
I will try to test and see if there is a way to make this work.

**Reply by Skrypt** [reply]
So by looking in Google :

zh-Hans, zh-CN = Chinese simplified. Mandarin Chinese, Simplified script, as used in China
zh-Hant, zh-TW = Chinese Taiwan, Traditional script

Correct me if I'm wrong.

**Reply by Skrypt** [reply]
No issue, go to sleep. Tomorrow is another day.

**Comment by snmin** [comment]
![image](https://user-images.githubusercontent.com/44972156/160519619-0bf8dffb-a119-4ffd-bbde-be917038bbe4.png)
Personally, I don't think it's a matter of localization of a single language, but it is invalid to switch any language after logging in. The valid part of translation is only the language selection box. I thought it was a browser problem before, cleaned the browser cache, and changed the browser. For example, if I choose to switch to Korean ko-KR now, it shows that the switch is successful, but it has not changed.

**Comment by snmin** [comment]
![image](https://user-images.githubusercontent.com/44972156/160521064-cb694444-dabf-4295-8225-21440e714940.png)

**Reply by Skrypt** [reply]
On the first request, it does that I think. After that, if you reload the entire page it will be translated. We can see that there is a translated notification.

**Reply by Skrypt** [reply]
Not every cultures are translated. For example here ko-KR translation has not been done yet. It depends on community contributions.

**Reply by Skrypt** [reply]
We have translations for "ja" culture.

**Reply by Skrypt** [reply]
Look at the /localization folder located in your main website project folder to see which translations exist.

**Reply by Skrypt** [reply]
One way to fix your issue is to copy one of the zh-XX folders and to rename it to zh in that folder. Then you will get chinese translations for every zh culture codes.

**Comment by snmin** [comment]
![image](https://user-images.githubusercontent.com/44972156/160521889-66a2801b-b9fa-485e-8f84-eb188ddcc018.png)
This is the interface that I choose to switch languages when I debug with Visual Studio locally. After switching, I can see the UI interface changes.

**Comment by snmin** [comment]
![image](https://user-images.githubusercontent.com/44972156/160522245-79dfb380-e9d5-4c76-9202-98080cc285e7.png)
This is what I tested with docker liunx. Selecting the same language and switching to save is invalid.

**Reply by Skrypt** [reply]
Ok what is your current browser prefered culture? Because it will display the Admin UI in the culture that your browser is set in order of priority. It is not a culture picker. You install cultures in this page but it's not a culture switcher. It will switch to your culture if your browser is set to that culture but else it won't even if you set the website default culture to zh for example. Because your browser is not accepting this culture or it is not set as prioritary in the culture order of your browser.

**Reply by snmin** [reply]
Chinese   zh_CN

**Reply by Skrypt** [reply]
So this culture page basically says: here are the cultures codes that I want to use to create content items.
The default culture is the fallback culture if someone navigates to your website and doesn't have any of the cultures that you have set in these settings in his browser prefered cultures/languages.

**Comment by snmin** [comment]
This is my docker linux with server deployment as test  
https://www.flameshine.cn/

**Reply by Skrypt** [reply]
Ok it fails as expected but we need to fix the zh-Hans culture thing in the source code.

**Reply by Skrypt** [reply]
![image](https://user-images.githubusercontent.com/3228637/160524011-d4bc9eaa-9d1e-429f-8988-77b6e521f953.png)

Here is a screenshot of the culture code as a reference to fix the issue.

**Reply by Skrypt** [reply]
Here is a screenshot of my Docker container for comparison : 

![image](https://user-images.githubusercontent.com/3228637/160524308-1b50b1db-bbdd-46af-89e7-87e499d1cc2b.png)

**Comment by snmin** [comment]
Yes, this test project uses the container released by you. I tried to export the localized language package in it. After opening it, I found that the language content is comprehensive and correct.

**Reply by snmin** [reply]
I just don't know why the switch fails in actual use.

**Reply by Skrypt** [reply]
Do you know how to upload folders/files inside a Docker container?

**Reply by snmin** [reply]
sudo docker cp *** orchardcms:/app/*

**Reply by Skrypt** [reply]
Because you could simply copy over the zh-CN and zh-TW folders and rename them to zh-Hans-CN and zh-Hant-TW
and them upload them in the docker container in the /Localization folder standing in the root of your website folder.

**Reply by snmin** [reply]
OK, I'll try.

**Reply by Skrypt** [reply]
And don't forget to restart the docker container afterward. It will restart the Orchard Core tenant and load these localizations.

**Reply by snmin** [reply]
![image](https://user-images.githubusercontent.com/44972156/160525408-484b48c8-31a8-4dae-a6b0-8567dd1ba8a9.png)
Can this be modified?

**Reply by Skrypt** [reply]
Yes, I think you can rename the folders directly in the container too.

**Reply by snmin** [reply]
I found no effect

**Reply by Skrypt** [reply]
Have you restarted the container?

**Reply by Skrypt** [reply]
Works now. I've set the proper cultures.

**Reply by snmin** [reply]
Yes, I should have forgotten to reboot. Thanks, it works now.
